File: ports/executor.go
package ports

import "context"

// ExecutionRequest is the input passed to an AgentExecutor.
// Agent properties are flattened here; the domain type core.AgentConfig
// is resolved by the bridge in cmd/ before the request reaches an executor.
type ExecutionRequest struct {
	SystemPrompt string
	Model        string
	Tools        []string
	Prompt       string
	RunID        string
	TaskID       string
}

// ExecutionResult is the output produced by an AgentExecutor.
type ExecutionResult struct {
	Output string
	Files  []string
}

// AgentExecutor runs a task against an AI backend and returns the result.
type AgentExecutor interface {
	Execute(ctx context.Context, request ExecutionRequest) (ExecutionResult, error)
}

File: stagehand/ports/executor.py
from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Callable


@dataclass
class ExecutionRequest:
    system_prompt: str = ""
    model: str = ""
    tools: list[str] = field(default_factory=list)
    prompt: str = ""
    run_id: str = ""
    task_id: str = ""


@dataclass
class ExecutionResult:
    output: str = ""
    files: list[str] = field(default_factory=list)
    data: Any = None


@dataclass
class ToolDefinition:
    """A custom tool that can be passed to an executor."""

    name: str
    description: str
    input_schema: dict[str, Any]
    handler: Callable[[dict[str, Any]], Any]


class AgentExecutor(ABC):
    """Port: runs a task against an AI backend and returns the result."""

    @abstractmethod
    async def execute(self, request: ExecutionRequest) -> ExecutionResult: ...
